test/upgrade.test.ts

```diff
@@ -0,0 +1,233 @@
+    const output: { type: 'log' | 'error'; message: unknown }[] = [];
+    return {
+      output,
+      log(this: void, message: unknown) {
+        output.push({ type: 'log', message });
+      },
+      logError(this: void, message: unknown) {
+        output.push({ type: 'error', message });
+      },
+    };
+  }
+
+  it('should report nothing to upgrade for an empty project', async () => {
+    const newPackagePath = join(fixturesRoot, 'new-package');
+    const { log, logError, output } = createMockOutput();
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+    await upgrade({
+      dryRun: true,
+      directoryPath: newPackagePath,
+      registryUrl: '',
+      createRegistry: () => createMockRegistry({}),
+      log,
+      logError,
+    });
+
+    expect(output).to.eql([
+      { type: 'log', message: 'Getting "latest" version for 0 dependencies...' },
+      { type: 'log', message: 'Nothing to upgrade.' },
+    ]);
+  });
+
+  it('should report nothing to upgrade when everything is up to date', async () => {
+    const newPackagePath = join(fixturesRoot, 'dependencies');
+    const { log, logError, output } = createMockOutput();
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+    await upgrade({
+      dryRun: true,
+      directoryPath: newPackagePath,
+      registryUrl: '',
+      createRegistry: () =>
+        createMockRegistry({
+          'package-a': { latest: '1.0.0' },
+          'package-b': { latest: '1.0.0' },
+          'package-c': { latest: '1.0.0' },
+        }),
+      log,
+      logError,
+    });
+
+    expect(output).to.eql([
+      { type: 'log', message: 'Getting "latest" version for 3 dependencies...' },
+      { type: 'log', message: 'Nothing to upgrade.' },
+    ]);
+  });
+
+  it('should report on dependencies with newer versions', async () => {
+    const newPackagePath = join(fixturesRoot, 'dependencies');
+    const { log, logError, output } = createMockOutput();
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+    await upgrade({
+      dryRun: true,
+      directoryPath: newPackagePath,
+      registryUrl: '',
+      createRegistry: () =>
+        createMockRegistry({
+          'package-a': { latest: '1.0.0' },
+          'package-b': { latest: '2.0.0' },
+          'package-c': { latest: '2.0.0' },
+        }),
+      log,
+      logError,
+    });
+
+    expect(output).to.eql([
+      { type: 'log', message: 'Getting "latest" version for 3 dependencies...' },
+      { type: 'log', message: 'Changes:' },
+      { type: 'log', message: '  package-b     ~1.0.0 -> ~2.0.0' },
+      { type: 'log', message: '  package-c     ^1.0.0 -> ^2.0.0' },
+    ]);
+  });
+
+  it('should force exact version to caret (opinionated)', async () => {
+    const newPackagePath = join(fixturesRoot, 'dependencies-exact');
+    const { log, logError, output } = createMockOutput();
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```

Comment:
need to remove this and enable experimental references in typescript-eslint

Reply:
added